### ISSUE TITLE
'output' flag in Measurement Group widget is only half respected

### ISSUE THREAD

**Issue by kklmn** [sardana-org/sardana#1123]
Hi!

We have a device in expchan that is not supposed to create data for data recorder. It requests a new image from LimaCCD. Other expchan elements read slices of that image as channels and ROIs. This 'counter' device produces nans in the DoorOutput with the default settings in the measurement group. If I select `false` in the `output` column, the nans in the door are gone together with their column header. Good. But the column of nans, and its header, is still present in the resulting data file. Not good. I consider this a bug.

Thanks!

**Comment by reszelaz** [comment]
Thanks @kklmn for reporting this!
Actually we don't use this feature at Alba, and I have no experience with it at all. When we were writing the documentation we [described it](https://sardana-controls.org/users/taurus/experimentconfiguration.html#experimental-channel-configuration) according to the current behavior, saying explicitelly about "printing" with the example of the output recorder.
But IMHO it does make a lot of sense to behave as you expect it i.e. not passing the value to any of the recorders. What could be even interpreted as not reading the value at all from the MacroServer level.
I encourage others to comment to see their point of view before classifying it as a bug. Thanks in advance!

**Comment by teresanunez** [comment]
Hi, I would also say that if the output is set to False the data should also not go to any of the recorders.
We also do not use this feature here, at least I am not aware of that.

**Comment by guifrecuni** [comment]
Hi, we designed the output flag to fulfill the request that is currently implemented and documented: hidden in spock, but included in the data files. If we want some element enabled and not recorded at all, neither for the output recorder nor for the rest of recorders, I'd name this flag 'recorded', not 'output'.

**Comment by reszelaz** [comment]
Thanks for @teresanunez and @guifrecuni for your comments. I classify it as a bug. The conclusion is that setting this flag should prevent from sending data to any of the recorders. I would already make the MeasurementGroup's `count()` method not reading the value.

If we go for renaming (output -> recorded), then we will also need to:
- rename the key in the measurement group's configuration dict (maintaining backwards compat, raising deprecation warning, etc.)
- provide new measurement group API methods, deprecating the old ones.
- rename the column in expconf and in macros (`get_meas_conf` and `set_meas_conf`).

It could perfectly be that renaming will take us more time than actually fixing the wrong behavior. Maybe we could make it incremental and start with a PR that just changes the behavior?

**Comment by teresanunez** [comment]
I agree with you. Thanks.

**Comment by cpascual** [comment]
> Maybe we could make it incremental and start with a PR that just changes the behavior?

I personally do not like the "incremental" solution and I do not agree with classifying this as a bug: as @guifrecuni  said, the **intended** (and aparently, documented) behaviour is exactly the current one. The reasoning behind it is that you may want to unclutter the output, but file size is "cheap" so it shouldn't bother you to have the whole info (incidentally, in the case of an array of NaNs, because of HDF5 transparent compression, the file size increase should be small).

Furthermore, the problem in this specific case comes, IMHO from an "abuse" of the system. In normal conditions, if you are neither interested in online visualization of  a channel during a scan nor in storing its data, then you simply *should not have the expchan in the measurement group*.  In this case, the "abuse" is to add this "dummy" channel to trigger other stuff... (which I think is a fine workaround, but it should not be the reason for changing a feature that is working as intended.

And in any case, if we want to change it, I think that it should be added as a different feature with a different name and then (optionally) deprecating the "output" one.

**Comment by cpascual** [comment]
because of my point of view that the "incremental" solution should not be followed, I add the `size:L` label

**Comment by reszelaz** [comment]
Thanks for the comment @cpascual, in general, you are right and I agree with you. However, I still have some doubts:

> The reasoning behind it is that you may want to unclutter the output, but file size is "cheap" so it shouldn't bother you to have the whole info (incidentally, in the case of an array of NaNs, because of HDF5 transparent compression, the file size increase should be small - asuming that you use hdf5 for storage).

You are right on the reason of having the output flag (uncluttering). I think that the same reason could apply to the SPEC data format users (still a [huge part](https://hackmd.io/@reszelaz/Bk53UScj_#/8/2) of the Sardana users), where visual inspection of the data is possible directly in the text editor. @kklmn are you one of those users? i.e. do you use the SPEC data format?

> Furthermore, the problem in this specific case comes, IMHO from an "abuse" of the system. In normal conditions, if you are neither interested in online visualization of a channel during a scan nor in storing its data, then you simply _should not have the expchan in the measurement group_. In this case, the "abuse" is to add this "dummy" channel to trigger other stuff... (which I think is a fine workaround, but it should not be the reason for changing a feature that is actually working as intended).

The scenario which @kklmn describes sounds like a _psuedo counters_ use case. So, even if in his case the experimental channel returns NaN, in a normal scenario we could have a _physical experimental channel_ returning a normal value (scalar, spectrum or image) and a _pseudo counter_ value returning the result of the calculation. And here comes a classical question, should we always store the raw data together with the processed data (pseudo counters). I imagine that both: _yes_ and _no_ answers are valid. So I would leave it to the user to decide. Since you can not have a pseudo counter in the measurement group without having the real counter (Sardana design), maybe Sardana needs what @kklmn asks for?

Also, due to the Sardana design, we are forced to use timer/monitor channels, which in most cases always return the same value of the integration time/monitor counts. You can not exclude them from the measurement group, but I imagine that the users may not need to repeat multiple columns with the same data.

> And in any case, if we want to change it, I think that it should be added as a different feature with a different name and then (optionally) deprecating the "output" one.

So, if both needs exist:
1. Unclutter the output e.g. Spock
2. Do not store certain experimental channels e.g. raw data, redundant timer/monitor, etc.

I agree with you to simply add another configuration. In this case, I wonder if selecting 2. should also imply 1.

**Comment by HEnquist** [comment]
I also think that the current behavior of "output" is correct. It's very useful for uncluttering the scan output. We often have quite large measurement groups where all columns simply can't be displayed reasonably.

**Comment by cpascual** [comment]
Also, just as an alternatie approach which could simplify implementation: if the problem is with ascii-based file recorders, one could always register a customized recorder that uses the output configuration or whatever other criteria (e.g. column name regexp matching) to skip writing certain channels.

IMHO the recorders plugin mechanism is flexible and easy to use enough to make this the most reasonable solution.

**Comment by kklmn** [comment]
> Also, due to the Sardana design, we are forced to use timer/monitor channels, which in most cases always return the same value of the integration time/monitor counts. You can not exclude them from the measurement group, but I imagine that the users may not need to repeat multiple columns with the same data.

This is absolutely true. When we use the same 4 signals by one or two Alba EMs, suddenly another timer column appears in the data file that we cannot remove and that spoils the data format in our analysis program.

The Measurement Group table model already has many columns that are irrelevant to us, why not adding another one for data writing?

I understand that for you guys the job is done when the file is written. That's why uncluttering the door output is more important for you than uncluttering column data files. For me (and for our beamline users) this is exactly the opposite and this is why I expected the output flag should be first of all for data writing. But it would just fine to have another flag that would do the job.

Thanks!

**Comment by dschick** [comment]
Hi @kklmn ,

I think the beauty of using SPEC or hdf5 files is to access the data columns by names and not by position in the data table.
This is why, I would suggest to update the analysis tools to respect the column names.
This is one of my first experiences from a Beamline, that the order and number of counters might always change even during a beamtime.
I started to develop a small [generic toolbox](github.com/dschick/pyEvalData) for such kind of data evaluation

However, I do have another use case for the `writing` column:
In case of pseudo counters from eg. 1D or 2D channels, such as ROIs, one might save the PC but not the source channels to save disk space.

Best

Daniel
